sw/device/silicon_creator/rom/e2e/rom_e2e_static_critical_test.c

```diff
@@ -0,0 +1,142 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_hmac.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/hmac_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/silicon_creator/lib/base/boot_measurements.h"
+#include "sw/device/silicon_creator/lib/base/sec_mmio.h"
+#include "sw/device/silicon_creator/lib/manifest_def.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+#define RUN_TEST(test_)                    \
+  LOG_INFO("Starting test " #test_ "..."); \
+  test_();                                 \
+  LOG_INFO("Finished test " #test_ ": ok.");
+
+#define CHECK_WORD_ALIGNED(addr_)                 \
+  CHECK((uintptr_t)addr_ % sizeof(uint32_t) == 0, \
+        #addr_ " must be word aligned");
+
+void boot_measurements_test(void) {
+  CHECK(kManifest.usage_constraints.selector_bits == 0,
+        "Selector bits must be 0");
+  const volatile char *manifest_start = (const volatile char *)&kManifest;
+  const char *manifest_end = (const char *)manifest_start + sizeof(manifest_t);
+  const volatile char *signed_region_start =
+      manifest_start + sizeof(sigverify_rsa_buffer_t);
+  const char *signed_region_end =
+      (const char *)manifest_start + kManifest.length;
+  size_t manifest_signed_region_size = manifest_end - signed_region_start;
+  size_t signed_region_size = signed_region_end - signed_region_start;
+  dif_hmac_t hmac;
+
+  CHECK_WORD_ALIGNED(manifest_start);
+  CHECK_WORD_ALIGNED(manifest_end);
+  CHECK_WORD_ALIGNED(signed_region_start);
+  CHECK_WORD_ALIGNED(signed_region_end);
+  CHECK_WORD_ALIGNED(manifest_signed_region_size);
+  CHECK_WORD_ALIGNED(signed_region_size);
+
+  LOG_INFO("Manifest: [0x%p, 0x%p)", manifest_start, manifest_end);
+  LOG_INFO("ROM_EXT signed region: [0x%p, 0x%p)", signed_region_start,
+           signed_region_end);
+  CHECK_DIF_OK(
+      dif_hmac_init(mmio_region_from_addr(TOP_EARLGREY_HMAC_BASE_ADDR), &hmac));
+  CHECK_DIF_OK(dif_hmac_mode_sha256_start(
+      &hmac, (dif_hmac_transaction_t){
+                 .digest_endianness = kDifHmacEndiannessLittle,
+                 .message_endianness = kDifHmacEndiannessLittle,
+             }));
+
+  // Copy the part of the manifest that's in the signed region to
+  // memory before pushing to hmac since it's volatile.
+  // Note: this array is larger than `manifest_signed_region_size` since VLAs
+  // are optional in C11.
+  char manifest_signed_region[sizeof(manifest_t)];
+  for (size_t i = 0; i < manifest_signed_region_size; ++i) {
+    manifest_signed_region[i] =
+        *((const volatile char *)signed_region_start + i);
```

Comment:
Question: I didn't follow the comment above. Is this needed because of the way we are setting permissions in ePMP?

Reply:
This is needed because the manifest is declared as `volatile` in `manifest_def.c`. We need it because the rest of the fields are filled by the tooling but this means we can't use `memcpy` or our utility functions because none of them expect `volatile` pointers. Resolving this thread but PLMK if this is not clear and I can work on the wording.